### ISSUE TITLE
[CI] Failure in SecurityCachePermissionTests

### ISSUE THREAD

**Issue by original-brownbear** [elastic/elasticsearch#44343]
This failure (https://scans.gradle.com/s/df24fs45kcg6k/) seems to point at a real issue in the SSL NIO transport.

This is what happened:

```

1> [2019-07-15T03:07:48,093][WARN ][o.e.t.TcpTransport       ] [node_s1] exception caught on transport layer [TcpNioSocketChannel{localAddress=null, remoteAddress=127.0.0.1/127.0.0.1:35148}], closing connection
--
1> java.lang.IllegalArgumentException: Cannot change mode after SSL traffic has started
1> 	at sun.security.ssl.SSLEngineImpl.setUseClientMode(SSLEngineImpl.java:1990) ~[?:?]
1> 	at org.elasticsearch.xpack.security.transport.nio.SSLDriver.init(SSLDriver.java:75) ~[main/:?]
1> 	at org.elasticsearch.xpack.security.transport.nio.SSLChannelContext.register(SSLChannelContext.java:60) ~[main/:?]
1> 	at org.elasticsearch.nio.EventHandler.handleRegistration(EventHandler.java:65) ~[elasticsearch-nio-7.4.0-SNAPSHOT.jar:7.4.0-SNAPSHOT]
1> 	at org.elasticsearch.nio.NioSelector.registerChannel(NioSelector.java:437) ~[elasticsearch-nio-7.4.0-SNAPSHOT.jar:7.4.0-SNAPSHOT]
1> 	at org.elasticsearch.nio.NioSelector.setUpNewChannels(NioSelector.java:429) ~[elasticsearch-nio-7.4.0-SNAPSHOT.jar:7.4.0-SNAPSHOT]
1> 	at org.elasticsearch.nio.NioSelector.preSelect(NioSelector.java:261) ~[elasticsearch-nio-7.4.0-SNAPSHOT.jar:7.4.0-SNAPSHOT]
1> 	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:155) ~[elasticsearch-nio-7.4.0-SNAPSHOT.jar:7.4.0-SNAPSHOT]
1> 	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:131) ~[elasticsearch-nio-7.4.0-SNAPSHOT.jar:7.4.0-SNAPSHOT]
1> 	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
1> [2019-07-15T03:07:48,125][INFO ][o.e.n.Node               ] [suite] stopped
1> [2019-07-15T03:07:48,125][INFO ][o.e.n.Node               ] [suite] closing ...
1> [2019-07-15T03:07:48,134][INFO ][o.e.n.Node               ] [suite] closed
2> jul 15, 2019 3:07:48 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
2> WARNING: Uncaught exception in thread: Thread[elasticsearch[node_s2][transport_worker][T#1],5,TGRP-SecurityCachePermissionTests]
2> java.lang.NullPointerException
2> 	at __randomizedtesting.SeedInfo.seed([AB646A88558E1481]:0)
2> 	at org.elasticsearch.nio.SelectionKeyUtils.isWriteInterested(SelectionKeyUtils.java:99)
2> 	at org.elasticsearch.nio.EventHandler.postHandling(EventHandler.java:184)
2> 	at org.elasticsearch.nio.NioSelector.writeToChannel(NioSelector.java:357)
2> 	at org.elasticsearch.nio.NioSelector.handleQueuedWrites(NioSelector.java:468)
2> 	at org.elasticsearch.nio.NioSelector.preSelect(NioSelector.java:262)
2> 	at org.elasticsearch.nio.NioSelector.singleLoop(NioSelector.java:155)
2> 	at org.elasticsearch.nio.NioSelector.runLoop(NioSelector.java:131)
2> 	at java.lang.Thread.run(Thread.java:748)


```

```
REPRODUCE WITH: ./gradlew :x-pack:plugin:security:test --tests "org.elasticsearch.integration.SecurityCachePermissionTests" -Dtests.seed=AB646A88558E1481 -Dtests.security.manager=true -Dtests.locale=en-US -Dtests.timezone=UTC -Dcompiler.java=12 -Druntime.java=8
```

Looking into it. Looks like we're trying to write to a broken channel that couldn't be properly connected and then NPE out because we don't have a selection key.

**Comment by original-brownbear** [comment]
This is sort of caused by #43653 I think. The underlying issue (running into an unexpected state for the SSL engine) probably isn't new but before that change the same NPE would happen in a `try`- catch all block and just be logged while now it bubbles up because we only use the selection key after the write. I'll see if I can track down the deeper issue with the SSL engine here.

**Comment by atorok** [comment]
another occurrence: https://elasticsearch-ci.elastic.co/view/Elasticsearch%207.x/job/elastic+elasticsearch+7.x+periodic/244/testReport/junit/org.elasticsearch.integration/DocumentLevelSecurityRandomTests/classMethod/

**Comment by original-brownbear** [comment]
Test failures would be fixed by https://github.com/elastic/elasticsearch/pull/44347,still looking into the underlying issue here

**Comment by original-brownbear** [comment]
Closing this as the test issue is fixed by #44347
